atomix/cluster/src/test/java/io/atomix/cluster/messaging/impl/NettyMessagingServiceTest.java

```diff
@@ -429,4 +432,214 @@ public void shouldNotBindToAdvertisedAddress() {
     assertThat(messagingService.bindingAddresses()).contains(bindingAddress);
     assertThat(messagingService.address()).isEqualTo(nonBindableAddress);
   }
+
+  @Test
+  public void testRemoteHandlerFailure() {
+    // given
+    final String exceptionMessage = "foo bar";
+    final RemoteHandlerFailure execptedException =
+        new MessagingException.RemoteHandlerFailure(exceptionMessage);
+
+    final BiFunction<Address, byte[], byte[]> handler =
+        (ep, data) -> {
+          throw new RuntimeException(exceptionMessage);
+        };
+
+    final String subject = nextSubject();
+    netty2.registerHandler(subject, handler, MoreExecutors.directExecutor());
+
+    // when
+    final CompletableFuture<byte[]> response =
+        netty1.sendAndReceive(address2, subject, "fail".getBytes());
+
+    // then
+    try {
+      response.join();
+      fail("MessagingException.RemoteHandlerFailure expected");
+    } catch (Exception e) {
+      final Throwable cause = e.getCause();
+      assertTrue(cause instanceof MessagingException.RemoteHandlerFailure);
+      assertTrue(execptedException.getMessage().equals(cause.getMessage()));
+    }
```

Comment:
:x: You'll notice that the forked Atomix code still uses some junit and Hamcrest assertions. For new code, or when you update that code, we always use AssertJ. So here, for example, you could use AssertJ's `Future`/`CompletableFuture` specific assertions, e.g.

```java
assertThat(response)
    .failsWithin(Duration.ofSeconds(30))
    .isInstanceOf(ExecutionException.class)
    .havingRootCause()
    .isInstanceOf(MessagingException.RemoteHandlerFailure)
    .havingMessage(cause.getMessage);
```

(I didn't test this code though, so maybe the method names aren't exactly right :sweat_smile:, but it should give you an idea)

Reply:
The same applies to the other tests below.

Reply:
Thanks for the hint. I applied the requested changes.